### PR TITLE
Fix some warnings

### PR DESCRIPTION
**Pull request by mehagar**
These aren't all of the warnings, just some of the easier ones to fix. The rest are mostly warnings about casting larger ints to smaller ints or similar.

### CODE REVIEW

**Comment by rajsite** [comment]
From 98 warnings to 42, nice 👍 

https://ci.appveyor.com/project/vireobot/vireosdk/build/1.0.537/messages

https://ci.appveyor.com/project/vireobot/vireosdk/build/1.0.538/messages

File: source/core/Events.cpp [hunk 1]
```diff
@@ -160,7 +160,7 @@ class EventQueueObject {
 
 typedef std::vector<EventQueueObject> EventQueueObjectVector;
 
-typedef Int32 EventOracleIndex;
+typedef IntIndex EventOracleIndex;
```
Comment (cglzaguilar) [inline source/core/Events.cpp]: Use size_t, instead of IntIndex and remove the static_casts, do the same for EventQueueID and remove the casts introduced for qID.

File: source/core/Events.cpp [hunk 2]
```diff
@@ -829,7 +829,7 @@ VIREO_FUNCTION_SIGNATURE2(UnregisterForEvents, RefNumVal, ErrorCluster)
 
 //------------------------------------------------------------
 struct WaitForEventArgs {
-    UInt32 *eventSpecIndex;
+    IntIndex *eventSpecIndex;
```
Comment (cglzaguilar) [inline source/core/Events.cpp]: Leave this as it was before (UInt32)

File: source/core/Events.cpp [hunk 3]
```diff
@@ -943,7 +943,7 @@ VIREO_FUNCTION_SIGNATUREV(WaitForEventsAndDispatch, WaitForEventsParamBlock)
                             eventRegRefnumPtr[regRefIndex-1].Type()->GetSubElement(0) : nullptr) : 0;
 
     for (Int32 inputTuple = 0; inputTuple < numTuples; ++inputTuple) {
-        UInt32 eventSpecIndex = *arguments[inputTuple].eventSpecIndex;
+        IntIndex eventSpecIndex = *arguments[inputTuple].eventSpecIndex;
```
Comment (cglzaguilar) [inline source/core/Events.cpp]: Leave this as it was before (UInt32)

File: source/core/Events.cpp [hunk 4]
```diff
@@ -943,7 +943,7 @@ VIREO_FUNCTION_SIGNATUREV(WaitForEventsAndDispatch, WaitForEventsParamBlock)
...
         if (eventSpecIndex >= esCount) {
```
Comment (cglzaguilar) [inline source/core/Events.cpp]: And either change the type of esCount to be UInt32 or cast here.

File: source/core/Events.cpp (hunk 1 above)
Comment (spathiwa) [reply source/core/Events.cpp]: Disagree, size_t is no good because:
  1) it's unsigned, but the sentinel kNotAnEventOracleIndex is -1.  Zero is a valid EventOracleIndex.
   We could change this (not much code relies on it), but this usage matches NCG.
 2) It's 8 bytes on native 64-bit platforms; I want it to be the same size everywhere.

File: source/core/NumericString.cpp [hunk 5]
```diff
@@ -488,7 +488,7 @@ void Format(SubString *format, Int32 count, StaticTypeAndData arguments[], Strin
                             if (tempDouble < 0) {
                             absDouble = 0.0 - tempDouble;
                             }
-                            exponent = floor(log10(absDouble));
+                            exponent = static_cast<Int32>(floor(log10(absDouble)));
```
Comment (spathiwa) [inline source/core/NumericString.cpp]: I think this is the sort of usage that reads better as a constructor rather than a cast:
`exponent = Int32(floor(log10(absDouble)));`

File: source/core/NumericString.cpp [hunk 6]
```diff
@@ -1315,7 +1315,7 @@ void DoubleScanString(StaticTypeAndData* argument, TypeRef argumentType, TempSta
         case 'B':
         case 'o': {
                 intValue = ScanIntBaseValues(formatChar, beginPointer, endPointer);
-                doubleValue = intValue;
+                doubleValue = static_cast<Double>(intValue);
```
Comment (spathiwa) [inline source/core/NumericString.cpp]: Ditto.

File: source/core/NumericString.cpp [hunk 7]
```diff
@@ -1349,7 +1349,7 @@ void DoubleScanString(StaticTypeAndData* argument, TypeRef argumentType, TempSta
             break;
         default: {
                 intValue = strtoll(beginPointer, endPointer, 10);
-                doubleValue = intValue;
+                doubleValue = static_cast<Double>(intValue);
```
Comment (spathiwa) [inline source/core/NumericString.cpp]: Ditto.

File: source/core/String.cpp [hunk 8]
```diff
@@ -508,15 +508,15 @@ VIREO_FUNCTION_SIGNATURE2(IsPrintable, StringRef, Boolean)
...
+        _Param(1) = isprint(c) != 0;
     }
     return _NextInstruction();
 }
 
 VIREO_FUNCTION_SIGNATURE2(IsPrintableInt, Int32, Boolean)
 {
     Utf8Char c = _Param(0);
-    _Param(1) = isprint(c);
+    _Param(1) = isprint(c) != 0;
```
Comment (spathiwa) [inline source/core/String.cpp]: Parentheses might help readability.

File: source/core/TDCodecVia.cpp [hunk 9]
```diff
@@ -1626,7 +1626,7 @@ void TDViaParser::PreParseClump(VIClump* viClump)
             break;
         } else {
             // Read its arguments.
-            tokenFound = _string.ReadSubexpressionToken(&token);
+            tokenFound = _string.ReadSubexpressionToken(&token) != TokenTraits_Unrecognized;
```
Comment (spathiwa) [inline source/core/TDCodecVia.cpp]: Ditto.

File: source/core/Events.cpp (hunk 1 above)
Comment (cglzaguilar) [reply source/core/Events.cpp]: Ok, I wasn't aware of the negative value.

File: source/core/Events.cpp (hunk 2 above)
Comment (mehagar) [reply source/core/Events.cpp]: See my next comment for why I would prefer to keep this as a signed int.

File: source/core/Events.cpp (hunk 4 above)
Comment (mehagar) [reply source/core/Events.cpp]: esCount is assigned from the value of a call to TypeCommon::SubElementCount, which returns an Int32. For the reason in my above comment I would prefer to keep this a signed int.

Comment (spathiwa) [reply source/core/Events.cpp]: This is an argument to WaitForEvents which comes from via code and is checked at run-time.
Changing it to Int32 would require also checking to see it's non-negative to validate it.
So this warning fix would slow run-time; that's not a good trade-off.

Instead, make esCount match.

`UInt32 esCount = UInt32(eventSpecType->SubElementCount());`

File: source/core/NumericString.cpp (hunk 5 above)
Comment (sanmut) [reply source/core/NumericString.cpp]: With modern C++, we could use uniform initialization (using braces instead of parentheses) by default. Where brace initialization throws compiler errors (due to narrowing), we could use C++ style casts. Is that acceptable to everyone?

File: source/core/Events.cpp [hunk 10]
```diff
@@ -184,7 +184,7 @@ class EventOracle {
         return _qObject[qID].DoneProcessingEvent();
     }
     void DeleteEventQueue(EventQueueID qID) {  // marks unallocated
-        if (qID < _qObject.size())
+        if (UInt32(qID) < _qObject.size())
```
[the remark this answers is not in the document]
Comment (mehagar) [reply source/core/Events.cpp]: For now I am going to revert this change.

[the remark this answers is not in the document]
Comment (spathiwa) [reply source/core/Events.cpp]: Using size_t(qID) would work and avoid a warning.

File: source/core/GenericFunctions.cpp [hunk 11]
```diff
@@ -722,7 +722,7 @@ static bool GetMinimumArrayDimensions(std::vector<TypedArrayCoreRef> arrays,
     IntIndex *array0Lengths = arrays[0]->DimensionLengths();
     for (int i = 0; i < array0Rank; i++) {
         IntIndex dimension0 = *(array0Lengths + i);
-        for (int j = 1; j < arrayCount; j++) {
+        for (int j = 1; j < Int32(arrayCount); j++) {
```
Comment (cglzaguilar) [inline source/core/GenericFunctions.cpp]: I think in this one we can make the type of j to be size_t, that way we can remove the cast

File: source/core/Timestamp.cpp [hunk 12]
```diff
@@ -141,7 +139,7 @@ namespace Vireo
         time_t t = _mkgmtime(&timeVal);
     #endif
     #endif
-        *this = Timestamp((Double)t + kStdDT1970re1904, fracSecs * 18446744073709551616.0);
+        *this = Timestamp((Double)t + kStdDT1970re1904 + fracSecs);
```
Comment (cglzaguilar) [inline source/core/Timestamp.cpp]: Please elaborate why this change is needed. This appears to be a change of behavior rather than a warning fix.

File: source/core/Events.cpp (hunk 1 above)
Comment (cglzaguilar) [reply source/core/Events.cpp]: I'm deferring review of changes to this file (Events.cpp) to @spathiwa

File: source/include/BuildConfig.h [hunk 13]
```diff
@@ -233,6 +233,7 @@ SDG
 #endif
 
 // TODO(PaulAustin): allow for thread locals on linux/unix
+#undef VIVM_THREAD_LOCAL
```
Comment (cglzaguilar) [inline source/include/BuildConfig.h]: Try the following instead:
#ifndef VIVM_THREAD_LOCAL
#define VIVM_THREAD_LOCAL
#endif

File: source/core/VirtualInstrument.cpp [hunk 14]
```diff
@@ -696,7 +696,7 @@ void ClumpParseState::InternalAddArgNeedingPatch(PatchInfo::PatchType patchType,
     // it should point to nullptr when checked if not yet resolved.
     _argPatches.push_back(_argCount);
     ++_argPatchCount;
-    if (_patchInfos.size() <= _patchInfoCount)
+    if (_patchInfos.size() <= UInt32(_patchInfoCount))
```
Comment (cglzaguilar) [inline source/core/VirtualInstrument.cpp]: Can we make the type of _patchInfoCount be size_t instead?

File: source/core/Timestamp.cpp (hunk 12 above)
Comment (mehagar) [reply source/core/Timestamp.cpp]: Yes, this was something I found while going through this file. It is the same fix as my previous timestamp precision fix, where we were losing precision by using the Timestamp constructor with two parameters. I am using the one with a Double parameter here instead, which preserves the precision. I don't know if this manifested as a bug or not.

File: source/core/GenericFunctions.cpp [hunk 15]
```diff
@@ -722,7 +722,7 @@ static bool GetMinimumArrayDimensions(std::vector<TypedArrayCoreRef> arrays,
     IntIndex *array0Lengths = arrays[0]->DimensionLengths();
     for (int i = 0; i < array0Rank; i++) {
         IntIndex dimension0 = *(array0Lengths + i);
-        for (int j = 1; j < Int32(arrayCount); j++) {
+        for (int j = 1; j < size_t(arrayCount); j++) {
```
Comment (cglzaguilar) [inline source/core/GenericFunctions.cpp]: Actually I was thinking something like this:
for (size_t j = 1; j<arrayCount; j++)
